### PR TITLE
Go 1.13  and 1.15 support with modules, updated examples and documentation (despite the branch name mentioning 1.14)

### PR DESCRIPTION
**Pull request by sciabarracom**
This PR includes support for Go 1.13 and 1.14 including many other improvements.
And supports Go Modules.

It should be functional, you can test it  but do not merge until I updated examples and documentation and removed older runtimes.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -16,7 +16,10 @@
 # limitations under the License.
 #
 -->
-# 1.16.0 (next release)
+# 1.16.0
```
Comment (rabbah) [inline CHANGES.md:19]:
```suggestion
# 1.16.0 (next release)
```

File: docs/BUILD.md [hunk 2]
```diff
@@ -43,10 +43,11 @@ To build the docker images, after compiling go proxy:
 
 This will build the images:
 
-* `action-golang-v1.11`: an image supporting  Go sources
-* `actionloop`: the base image, supporting generic executables ans shell script
+* `action-golang-v1.13`: an image supporting Go 1.13 sources (does expect an ack)
+* `action-golang-v1.14`: an image supporting Go 1.14 sources (does expect an ack)
```
Comment (style95) [inline docs/BUILD.md:47]: Should we mention v1.15 as well?

File: docs/ACTION.md [hunk 3]
```diff
@@ -22,7 +22,7 @@
 
 ## How to write Go Actions
 
-The `action-golang-v1.11` runtime can execute actions written in the Go programming language in OpenWhisk, either precompiled binary or compiling sources on the fly.
+The `action-golang-v1.14` runtime can execute actions written in the Go programming language in OpenWhisk, either precompiled binary or compiling sources on the fly.
```
Comment (style95) [inline docs/ACTION.md:25]: Would it be better to change this to v1.15?

File: docs/DEPLOY.md [hunk 4]
```diff
@@ -49,120 +49,41 @@ The runtime `action-golang-v1.11` accepts:
...
-
-Using this option, the GOPATH will be set to the parent directory of your `src` folder and you will not have errors in your imports.
+Check the example: `package-main` and `module-main` and look for the format of the `go.mod` files.
 
 <a name="precompile"/>
-
 ## Precompiling Go Sources Offline
 
-Compiling sources on the image can take some time when the images is initialized. You can speed up precompiling the sources using the image `action-golang-v1.11` as an offline compiler. You need `docker` for doing that.
+Compiling sources on the image can take some time when the images is initialized. You can speed up precompiling the sources using the image `action-golang-v1.14` as an offline compiler. You need `docker` for doing that.
```
Comment (style95) [inline docs/DEPLOY.md:73]: Same with above.

File: examples/EXAMPLES.md [hunk 5]
```diff
@@ -22,29 +22,24 @@
 This is a collection of examples.
 Tested on:
 
-- dep version 0.5.0 (check the version `dep version`)
-- Go version 1.11.1
+- Go version 1.14.6
```
Comment (style95) [inline examples/EXAMPLES.md:25]: a nit

File: examples/module-main/go.mod [hunk 6]
```diff
@@ -0,0 +1,5 @@
+module action
+
+go 1.14
```
Comment (style95) [inline examples/module-main/go.mod:3]: a nit

File: examples/package-main/go.mod [hunk 7]
```diff
@@ -0,0 +1,7 @@
+module action
+
+go 1.14
```
Comment (style95) [inline examples/package-main/go.mod:3]: a nit

File: examples/package-main/hello/go.mod [hunk 8]
```diff
@@ -0,0 +1,3 @@
+module hello
+
+go 1.14
```
Comment (style95) [inline examples/package-main/hello/go.mod:3]: a nit

File: golang1.14/Dockerfile [hunk 9]
```diff
@@ -14,27 +14,28 @@
...
-	docker kill under-test
-
-bash: test.lua
-	docker run -d --name under-test --rm -p 8080:8080 $(IMG2)
-	bash init.sh main.sh
-	wrk -t1 -c1 -stest.lua http://localhost:8080/run
-	docker kill under-test
-
-.PHONY: all golang bash
+FROM golang:1.13.8
```
Comment (style95) [inline golang1.14/Dockerfile:17]: Should this be some 1.14.x version?

**Review by style95 — COMMENTED** [review]
If we start supporting go-1.15, how about updating documents accordingly too?

File: golang1.14/Dockerfile (hunk 9 above)
Comment (sciabarracom) [reply golang1.14/Dockerfile:17]: ouch! How I did miss this?

**Comment by mrutkows** [comment]
@sciabarracom  I have been warned away from using go 1.14.x in any production setting due to EINTR issues...  the advice is to skip from 1.13 to 1.15 (where the comm. addressed regressions in sys libs.) which thankfully was just released a couple of weeks ago.

**Comment by mrutkows** [comment]
@sciabarra The issue is ack'ed in the 1.14 release notes:
https://golang.org/doc/go1.14

Specifically:
"Goroutines are now asynchronously preemptible. As a result, loops without function calls no longer potentially deadlock the scheduler or significantly delay garbage collection. This is supported on all platforms except windows/arm, darwin/arm, js/wasm, and plan9/*.

A consequence of the implementation of preemption is that on Unix systems, including Linux and macOS systems, programs built with Go 1.14 will receive more signals than programs built with earlier releases. This means that programs that use packages like syscall or golang.org/x/sys/unix will see more slow system calls fail with EINTR errors. Those programs will have to handle those errors in some way, most likely looping to try the system call again. For more information about this see man 7 signal for Linux systems or similar documentation for other systems. "

Under production loads, many sys. level calls will fail because they are not coded to be interrupt tolerant (i.e., code their own retries on failure) resulting higher order errors.

**Comment by sciabarracom** [comment]
Yes but what we do? Do we get rid of all the runtimes older than 1.15? Or just skip 1.14? 

-- 
  Michele Sciabarra
  michele@sciabarra.com



----- Original message -----

**Comment by mrutkows** [comment]
Hi Michele,

You must have missed the comment I made previously above the one you 
copied below.  It reads:

"@sciabarracom I have been warned away from using go 1.14.x in any 
production setting due to EINTR issues... the advice is to skip from 1.13 
to 1.15 (where the comm. addressed regressions in sys libs.) which 
thankfully was just released a couple of weeks ago."

**Comment by sciabarra** [comment]
I was asking if we have to keep 1.13

-- 
  Michele Sciabarra
  github@sciabarra.com



----- Original message -----

**Comment by mrutkows** [comment]
Had not formulated an opinion on 1.13 strongly until now; however, with 
1.15 being so new and 1.14 being skipped, the 1.13 revisions do include 
some interesting and sig. fixes and additions including many security and 
crypto fixes.  Could see some demand the 1.13 runtimes in addition as my 
view it is the last stable minor revision that still includes support for 
older CPUs and 32-bit as well. 

If it were me, I would mint a 1.13.15 runtime and then work on 1.15.x and 
if someone wants a 1.14 for some reason they can add it.

**Comment by sciabarracom** [comment]
I actually the runtimes are ready  with a PR pending. There is still one test failing because tests does not include a proper go.mod but it is fault of the tests not of the runtime. Will fix the test, leaving 1.13 and 1.15. I was asking in order to complete the PR.


-- 
  Michele Sciabarra
  michele@sciabarra.com



----- Original message -----

**Pull request merged**